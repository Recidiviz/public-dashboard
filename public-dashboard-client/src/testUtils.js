// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import fs from "fs";
import path from "path";
import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components/macro";
import { THEME } from "./theme";
import { InfoPanelProvider } from "./info-panel";

// eslint-disable-next-line react/prop-types
const GlobalWrapper = ({ children }) => {
  // include globally expected Context providers or other required wrappers here
  return (
    <ThemeProvider theme={THEME}>
      <InfoPanelProvider>{children}</InfoPanelProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalWrapper, ...options });

// re-export everything from testing-library
export * from "@testing-library/react";

// override render method
export { customRender as render };

// provide original render method as a fallback if needed
export { render as renderUnwrapped };

// retrieve a data fixture from spotlight-api
export function getDataFixture(filename) {
  const filePath = path.resolve(
    __dirname,
    `../../spotlight-api/core/demo_data/${filename}`
  );
  const stringContents = fs.readFileSync(filePath).toString();

  // copypasta from spotlight-api/core/metricsApi for transforming JSONLines
  if (!stringContents || stringContents.length === 0) {
    return null;
  }
  const jsonObject = [];
  const splitStrings = stringContents.split("\n");
  splitStrings.forEach((line) => {
    if (line) {
      jsonObject.push(JSON.parse(line));
    }
  });

  return jsonObject;
}
