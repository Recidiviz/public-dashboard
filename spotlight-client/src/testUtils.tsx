// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { LocationProvider } from "@reach/router";
import { render } from "@testing-library/react";
import React from "react";
import { autorun } from "mobx";
import waitForLocalhost from "wait-for-localhost";

// eslint-disable-next-line import/prefer-default-export
export function waitForTestServer(): Promise<void> {
  return waitForLocalhost({ path: "/health", port: 3002 });
}

/**
 * Convenience method to run an immediate, one-time reactive effect
 */
export function reactImmediately(effect: () => void): void {
  // this will call the effect function immediately,
  // and then immediately call the disposer to tear down the reaction
  autorun(effect)();
}

const LocationContextWrapper: React.FC = ({ children }) => {
  return <LocationProvider>{children}</LocationProvider>;
};

/**
 * Convenience method for rendering components that use @reach/router hooks
 * in a LocationContext to prevent rendering errors.
 */
export const renderWithRouter = (
  ui: React.ReactElement
): ReturnType<typeof render> => {
  return render(ui, { wrapper: LocationContextWrapper });
};
