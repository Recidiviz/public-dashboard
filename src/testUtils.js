import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { THEME } from "./constants";

// eslint-disable-next-line react/prop-types
const GlobalWrapper = ({ children }) => {
  // include globally expected Context providers or other required wrappers here
  return <ThemeProvider theme={THEME}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalWrapper, ...options });

// re-export everything from testing-library
export * from "@testing-library/react";

// override render method
export { customRender as render };

// provide original render method as a fallback if needed
export { render as renderUnwrapped };
