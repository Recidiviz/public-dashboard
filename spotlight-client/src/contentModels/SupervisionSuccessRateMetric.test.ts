// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { csvParse } from "d3-dsv";
import downloadjs from "downloadjs";
import { advanceTo, clear } from "jest-date-mock";
import JsZip from "jszip";
import { runInAction, when } from "mobx";
import { stripHtml } from "string-strip-html";
import {
  fetchAndTransformMetric,
  SupervisionSuccessRateMonthlyRecord,
} from "../metricsApi";
import { reactImmediately } from "../testUtils";
import contentFixture from "./__fixtures__/tenant_content_exhaustive";
import createMetricMapping from "./createMetricMapping";
import SupervisionSuccessRateMetric from "./SupervisionSuccessRateMetric";

jest.mock("../metricsApi", () => {
  const actualModule = jest.requireActual("../metricsApi");
  const mockFetch = jest.fn(actualModule.fetchAndTransformMetric);
  return { ...actualModule, fetchAndTransformMetric: mockFetch };
});

const mockedFetchAndTransformMetric = fetchAndTransformMetric as jest.MockedFunction<
  typeof fetchAndTransformMetric
>;

jest.mock("downloadjs");
const downloadjsMock = downloadjs as jest.MockedFunction<typeof downloadjs>;

const testTenantId = "US_ND";
const testMetricId = "ProbationTerminationsHistorical";
const testMetadataMapping = {
  [testMetricId]: contentFixture.metrics[testMetricId],
};

function getTestMetric() {
  return createMetricMapping({
    localityLabelMapping: contentFixture.localities,
    metadataMapping: testMetadataMapping,
    tenantId: testTenantId,
    demographicFilter: contentFixture.demographicCategories,
  }).get(testMetricId) as SupervisionSuccessRateMetric;
}

async function getPopulatedMetric() {
  const metric = getTestMetric();

  metric.hydrate();

  await when(() => !metric.isLoading && metric.error === undefined);

  return metric;
}

beforeEach(() => {
  // last month in data fixture
  advanceTo(new Date(2020, 6, 2));
});

afterEach(() => {
  clear();
});

describe("cohort data", () => {
  test("total", async () => {
    const metric = await getPopulatedMetric();

    reactImmediately(() => {
      expect(metric.cohortRecords).toMatchSnapshot();
      expect(metric.cohortRecords?.length).toBe(36);
    });

    expect.hasAssertions();
  });

  test("filtered by locality", async () => {
    const metric = await getPopulatedMetric();

    runInAction(() => {
      metric.localityId = contentFixture.localities.Probation.entries[0].id;
    });

    reactImmediately(() => {
      expect(metric.cohortRecords).toMatchSnapshot();
    });

    expect.hasAssertions();
  });

  test("not filtered by demographics", async () => {
    const metric = await getPopulatedMetric();
    const initialCohortRecords: typeof metric.cohortRecords = [];

    reactImmediately(() => {
      const totalRecords = metric.cohortRecords;
      if (totalRecords && totalRecords.length) {
        initialCohortRecords.push(...totalRecords);
      }
      expect(initialCohortRecords.length).toBeGreaterThan(0);
    });

    runInAction(() => {
      metric.demographicView = "raceOrEthnicity";
    });

    reactImmediately(() => {
      expect(metric.cohortRecords).toEqual(initialCohortRecords);
    });

    expect.hasAssertions();
  });

  test("imputes missing cohorts", async () => {
    mockedFetchAndTransformMetric.mockResolvedValueOnce([
      {
        year: 2019,
        month: 9,
        district: "ALL",
        rateNumerator: 2,
        rateDenominator: 2,
        rate: 1,
      },
    ]);

    const metric = await getPopulatedMetric();

    reactImmediately(() => {
      const cohortRecords = metric.cohortRecords as SupervisionSuccessRateMonthlyRecord[];
      expect(cohortRecords.length).toEqual(36);
      cohortRecords.forEach((record) => {
        if (record.year !== 2019 && record.month !== 9)
          expect(record).toEqual(
            expect.objectContaining({
              rate: 0,
              rateDenominator: 0,
              rateNumerator: 0,
            })
          );
      });

      // does not include current month, as set via jest date mock
      expect(cohortRecords).toEqual(
        expect.not.arrayContaining([
          expect.objectContaining({ year: 2020, month: 7 }),
        ])
      );
      // as a result, the start should be shifted back a month
      expect(cohortRecords).toEqual(
        expect.arrayContaining([
          expect.objectContaining({ year: 2017, month: 7 }),
        ])
      );
    });

    expect.hasAssertions();
  });

  test("imputes missing cohorts with current month", async () => {
    mockedFetchAndTransformMetric.mockResolvedValueOnce([
      {
        year: 2020,
        month: 7,
        district: "ALL",
        rateNumerator: 2,
        rateDenominator: 2,
        rate: 1,
      },
    ]);

    const metric = await getPopulatedMetric();

    reactImmediately(() => {
      const cohortRecords = metric.cohortRecords as SupervisionSuccessRateMonthlyRecord[];
      expect(cohortRecords.length).toEqual(36);
      cohortRecords.forEach((record) => {
        if (record.year !== 2020 && record.month !== 7)
          expect(record).toEqual(
            expect.objectContaining({
              rate: 0,
              rateDenominator: 0,
              rateNumerator: 0,
            })
          );
      });
      // includes current month, as set via jest date mock
      expect(cohortRecords).toEqual(
        expect.arrayContaining([
          expect.objectContaining({ year: 2020, month: 7 }),
        ])
      );
      // as a result, the start is shifted forward a month
      expect(cohortRecords).toEqual(
        expect.not.arrayContaining([
          expect.objectContaining({ year: 2017, month: 7 }),
        ])
      );
    });

    expect.hasAssertions();
  });
});

// This whole test suite fails because of this section. See the issue #589
//
// describe("demographic data", () => {
//   test("total", async () => {
//     const metric = await getPopulatedMetric();

//     reactImmediately(() => {
//       expect(metric.demographicRecords).toMatchSnapshot();
//     });

//     expect.hasAssertions();
//   });

//   test("filtered by locality", async () => {
//     const metric = await getPopulatedMetric();

//     runInAction(() => {
//       metric.localityId = contentFixture.localities.Probation.entries[0].id;
//     });

//     reactImmediately(() => {
//       expect(metric.demographicRecords).toMatchSnapshot();
//     });

//     expect.hasAssertions();
//   });

//   test.each([["raceOrEthnicity"], ["gender"], ["ageBucket"]] as [
//     Exclude<DemographicView, "nofilter">
//   ][])("%s data", async (demographicView) => {
//     const metric = await getPopulatedMetric();

//     runInAction(() => {
//       metric.demographicView = demographicView;
//     });

//     reactImmediately(() => {
//       expect(metric.demographicRecords).toMatchSnapshot();
//     });

//     expect.hasAssertions();
//   });
// });

test("data download", async (done) => {
  expect.hasAssertions();
  const metric = await getPopulatedMetric();

  await metric.download();

  expect(downloadjsMock).toHaveBeenCalled();

  const [content, filename] = downloadjsMock.mock.calls[0];
  const prefix = `${testTenantId} ${testMetricId} data`;
  expect(filename).toBe(`${prefix}.zip`);

  // if we read the data in the zip file we should be able to reverse it
  // into something resembling metric.allRecords; we will spot-check
  // the downloaded file by verifying that the first record matches the source
  const zip = await JsZip.loadAsync(content);
  reactImmediately(async () => {
    // expecting two files and their respective data sources
    const expectedFiles = [
      {
        name: `${prefix}/historical data.csv`,
        source: metric.allCohortRecords,
      },
      {
        name: `${prefix}/demographic aggregate data.csv`,
        source: metric.allDemographicRecords,
      },
    ];

    await Promise.all(
      expectedFiles.map(async ({ name, source }) => {
        const csvContents = await zip.file(name)?.async("string");

        if (csvContents) {
          const recordsFromCsv = csvParse(csvContents);

          // rather than try to re-typecast the CSV data,
          // we'll cast the real record value to strings for comparison
          const expectedRecord: Record<string, string> = {};

          // in practice, source should not be undefined once we've gotten this far
          Object.entries((source || [])[0]).forEach(([key, value]) => {
            expectedRecord[key] = String(value);
          });
          expect(recordsFromCsv[0]).toEqual(expectedRecord);
        } else {
          throw new Error("unable to retrieve CSV data from zip file");
        }
      })
    );

    const readmeContents = await zip
      .file(`${prefix}/README.txt`)
      ?.async("string");
    // the file in the archive is plain text but methodology can contain HTML tags
    expect(readmeContents).toBe(stripHtml(metric.methodology).result);

    done();
  });
});

test("no unknowns", async () => {
  const metric = getTestMetric();
  await metric.hydrate();

  reactImmediately(() => {
    expect(metric.unknowns).toBeUndefined();
  });

  expect.hasAssertions();
});

test("report unknowns", (done) => {
  const metric = getTestMetric();

  // cohort data, which we don't care about here
  mockedFetchAndTransformMetric.mockResolvedValueOnce([]);
  // demographic data
  mockedFetchAndTransformMetric.mockResolvedValueOnce([
    {
      locality: "ALL",
      raceOrEthnicity: "ALL",
      gender: "ALL",
      ageBucket: "EXTERNAL_UNKNOWN",
      rateNumerator: 1,
      rateDenominator: 2,
      rate: 0.5,
    },
    {
      locality: "test1",
      raceOrEthnicity: "ALL",
      gender: "ALL",
      ageBucket: "EXTERNAL_UNKNOWN",
      rateNumerator: 0,
      rateDenominator: 1,
      rate: 0,
    },
    {
      locality: "test2",
      raceOrEthnicity: "ALL",
      gender: "ALL",
      ageBucket: "EXTERNAL_UNKNOWN",
      rateNumerator: 1,
      rateDenominator: 1,
      rate: 1,
    },
  ]);

  metric.hydrate();

  when(
    () => metric.unknowns !== undefined,
    () => {
      expect(metric.unknowns).toEqual({
        raceOrEthnicity: 0,
        gender: 0,
        ageBucket: 2,
      });

      runInAction(() => {
        metric.localityId = "test1";
      });
      expect(metric.unknowns).toEqual({
        raceOrEthnicity: 0,
        gender: 0,
        ageBucket: 1,
      });

      // this filter should be ignored
      runInAction(() => {
        metric.demographicView = "raceOrEthnicity";
      });
      expect(metric.unknowns).toEqual({
        raceOrEthnicity: 0,
        gender: 0,
        ageBucket: 1,
      });
      done();
    }
  );
});
