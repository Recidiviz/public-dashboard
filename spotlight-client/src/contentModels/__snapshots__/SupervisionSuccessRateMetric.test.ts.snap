// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cohort data filtered by locality 1`] = `
Array [
  Object {
    "label": "Aug 2017",
    "locality": "NORTHEAST",
    "month": 8,
    "rate": 0.3448275862068966,
    "rateDenominator": 29,
    "rateNumerator": 10,
    "year": 2017,
  },
  Object {
    "label": "Sep 2017",
    "locality": "NORTHEAST",
    "month": 9,
    "rate": 0.47368421052631576,
    "rateDenominator": 19,
    "rateNumerator": 9,
    "year": 2017,
  },
  Object {
    "label": "Oct 2017",
    "locality": "NORTHEAST",
    "month": 10,
    "rate": 0.5555555555555556,
    "rateDenominator": 27,
    "rateNumerator": 15,
    "year": 2017,
  },
  Object {
    "label": "Nov 2017",
    "locality": "NORTHEAST",
    "month": 11,
    "rate": 0.24,
    "rateDenominator": 25,
    "rateNumerator": 6,
    "year": 2017,
  },
  Object {
    "label": "Dec 2017",
    "locality": "NORTHEAST",
    "month": 12,
    "rate": 0.30434782608695654,
    "rateDenominator": 23,
    "rateNumerator": 7,
    "year": 2017,
  },
  Object {
    "label": "Jan 2018",
    "locality": "NORTHEAST",
    "month": 1,
    "rate": 0.3023255813953488,
    "rateDenominator": 43,
    "rateNumerator": 13,
    "year": 2018,
  },
  Object {
    "label": "Feb 2018",
    "locality": "NORTHEAST",
    "month": 2,
    "rate": 0.6551724137931034,
    "rateDenominator": 29,
    "rateNumerator": 19,
    "year": 2018,
  },
  Object {
    "label": "Mar 2018",
    "locality": "NORTHEAST",
    "month": 3,
    "rate": 0.52,
    "rateDenominator": 25,
    "rateNumerator": 13,
    "year": 2018,
  },
  Object {
    "label": "Apr 2018",
    "locality": "NORTHEAST",
    "month": 4,
    "rate": 0.7916666666666666,
    "rateDenominator": 24,
    "rateNumerator": 19,
    "year": 2018,
  },
  Object {
    "label": "May 2018",
    "locality": "NORTHEAST",
    "month": 5,
    "rate": 0.3333333333333333,
    "rateDenominator": 42,
    "rateNumerator": 14,
    "year": 2018,
  },
  Object {
    "label": "Jun 2018",
    "locality": "NORTHEAST",
    "month": 6,
    "rate": 0.45161290322580644,
    "rateDenominator": 31,
    "rateNumerator": 14,
    "year": 2018,
  },
  Object {
    "label": "Jul 2018",
    "locality": "NORTHEAST",
    "month": 7,
    "rate": 0.42857142857142855,
    "rateDenominator": 21,
    "rateNumerator": 9,
    "year": 2018,
  },
  Object {
    "label": "Aug 2018",
    "locality": "NORTHEAST",
    "month": 8,
    "rate": 0.4,
    "rateDenominator": 10,
    "rateNumerator": 4,
    "year": 2018,
  },
  Object {
    "label": "Sep 2018",
    "locality": "NORTHEAST",
    "month": 9,
    "rate": 0.5263157894736842,
    "rateDenominator": 38,
    "rateNumerator": 20,
    "year": 2018,
  },
  Object {
    "label": "Oct 2018",
    "locality": "NORTHEAST",
    "month": 10,
    "rate": 0.4444444444444444,
    "rateDenominator": 18,
    "rateNumerator": 8,
    "year": 2018,
  },
  Object {
    "label": "Nov 2018",
    "locality": "NORTHEAST",
    "month": 11,
    "rate": 0.45,
    "rateDenominator": 40,
    "rateNumerator": 18,
    "year": 2018,
  },
  Object {
    "label": "Dec 2018",
    "locality": "NORTHEAST",
    "month": 12,
    "rate": 0.5277777777777778,
    "rateDenominator": 36,
    "rateNumerator": 19,
    "year": 2018,
  },
  Object {
    "label": "Jan 2019",
    "locality": "NORTHEAST",
    "month": 1,
    "rate": 0.47058823529411764,
    "rateDenominator": 34,
    "rateNumerator": 16,
    "year": 2019,
  },
  Object {
    "label": "Feb 2019",
    "locality": "NORTHEAST",
    "month": 2,
    "rate": 0.391304347826087,
    "rateDenominator": 23,
    "rateNumerator": 9,
    "year": 2019,
  },
  Object {
    "label": "Mar 2019",
    "locality": "NORTHEAST",
    "month": 3,
    "rate": 0.4827586206896552,
    "rateDenominator": 29,
    "rateNumerator": 14,
    "year": 2019,
  },
  Object {
    "label": "Apr 2019",
    "locality": "NORTHEAST",
    "month": 4,
    "rate": 0.4583333333333333,
    "rateDenominator": 24,
    "rateNumerator": 11,
    "year": 2019,
  },
  Object {
    "label": "May 2019",
    "locality": "NORTHEAST",
    "month": 5,
    "rate": 0.325,
    "rateDenominator": 40,
    "rateNumerator": 13,
    "year": 2019,
  },
  Object {
    "label": "Jun 2019",
    "locality": "NORTHEAST",
    "month": 6,
    "rate": 0.4117647058823529,
    "rateDenominator": 17,
    "rateNumerator": 7,
    "year": 2019,
  },
  Object {
    "label": "Jul 2019",
    "locality": "NORTHEAST",
    "month": 7,
    "rate": 0.3939393939393939,
    "rateDenominator": 33,
    "rateNumerator": 13,
    "year": 2019,
  },
  Object {
    "label": "Aug 2019",
    "locality": "NORTHEAST",
    "month": 8,
    "rate": 0.45454545454545453,
    "rateDenominator": 33,
    "rateNumerator": 15,
    "year": 2019,
  },
  Object {
    "label": "Sep 2019",
    "locality": "NORTHEAST",
    "month": 9,
    "rate": 0.38461538461538464,
    "rateDenominator": 26,
    "rateNumerator": 10,
    "year": 2019,
  },
  Object {
    "label": "Oct 2019",
    "locality": "NORTHEAST",
    "month": 10,
    "rate": 0.4473684210526316,
    "rateDenominator": 38,
    "rateNumerator": 17,
    "year": 2019,
  },
  Object {
    "label": "Nov 2019",
    "locality": "NORTHEAST",
    "month": 11,
    "rate": 0.32,
    "rateDenominator": 25,
    "rateNumerator": 8,
    "year": 2019,
  },
  Object {
    "label": "Dec 2019",
    "locality": "NORTHEAST",
    "month": 12,
    "rate": 0.4,
    "rateDenominator": 25,
    "rateNumerator": 10,
    "year": 2019,
  },
  Object {
    "label": "Jan 2020",
    "locality": "NORTHEAST",
    "month": 1,
    "rate": 0.35,
    "rateDenominator": 40,
    "rateNumerator": 14,
    "year": 2020,
  },
  Object {
    "label": "Feb 2020",
    "locality": "NORTHEAST",
    "month": 2,
    "rate": 0.45454545454545453,
    "rateDenominator": 22,
    "rateNumerator": 10,
    "year": 2020,
  },
  Object {
    "label": "Mar 2020",
    "locality": "NORTHEAST",
    "month": 3,
    "rate": 0.3333333333333333,
    "rateDenominator": 24,
    "rateNumerator": 8,
    "year": 2020,
  },
  Object {
    "label": "Apr 2020",
    "locality": "NORTHEAST",
    "month": 4,
    "rate": 0.41935483870967744,
    "rateDenominator": 31,
    "rateNumerator": 13,
    "year": 2020,
  },
  Object {
    "label": "May 2020",
    "locality": "NORTHEAST",
    "month": 5,
    "rate": 0.36666666666666664,
    "rateDenominator": 30,
    "rateNumerator": 11,
    "year": 2020,
  },
  Object {
    "label": "Jun 2020",
    "locality": "NORTHEAST",
    "month": 6,
    "rate": 0.3333333333333333,
    "rateDenominator": 9,
    "rateNumerator": 3,
    "year": 2020,
  },
  Object {
    "label": "Jul 2020",
    "locality": "NORTHEAST",
    "month": 7,
    "rate": 0.125,
    "rateDenominator": 8,
    "rateNumerator": 1,
    "year": 2020,
  },
]
`;

exports[`cohort data total 1`] = `
Array [
  Object {
    "label": "Aug 2017",
    "locality": "ALL",
    "month": 8,
    "rate": 0.5205479452054794,
    "rateDenominator": 365,
    "rateNumerator": 190,
    "year": 2017,
  },
  Object {
    "label": "Sep 2017",
    "locality": "ALL",
    "month": 9,
    "rate": 0.5802469135802469,
    "rateDenominator": 324,
    "rateNumerator": 188,
    "year": 2017,
  },
  Object {
    "label": "Oct 2017",
    "locality": "ALL",
    "month": 10,
    "rate": 0.536723163841808,
    "rateDenominator": 354,
    "rateNumerator": 190,
    "year": 2017,
  },
  Object {
    "label": "Nov 2017",
    "locality": "ALL",
    "month": 11,
    "rate": 0.5056179775280899,
    "rateDenominator": 445,
    "rateNumerator": 225,
    "year": 2017,
  },
  Object {
    "label": "Dec 2017",
    "locality": "ALL",
    "month": 12,
    "rate": 0.49056603773584906,
    "rateDenominator": 318,
    "rateNumerator": 156,
    "year": 2017,
  },
  Object {
    "label": "Jan 2018",
    "locality": "ALL",
    "month": 1,
    "rate": 0.5348837209302325,
    "rateDenominator": 559,
    "rateNumerator": 299,
    "year": 2018,
  },
  Object {
    "label": "Feb 2018",
    "locality": "ALL",
    "month": 2,
    "rate": 0.5067961165048543,
    "rateDenominator": 515,
    "rateNumerator": 261,
    "year": 2018,
  },
  Object {
    "label": "Mar 2018",
    "locality": "ALL",
    "month": 3,
    "rate": 0.5607476635514018,
    "rateDenominator": 428,
    "rateNumerator": 240,
    "year": 2018,
  },
  Object {
    "label": "Apr 2018",
    "locality": "ALL",
    "month": 4,
    "rate": 0.5603217158176944,
    "rateDenominator": 373,
    "rateNumerator": 209,
    "year": 2018,
  },
  Object {
    "label": "May 2018",
    "locality": "ALL",
    "month": 5,
    "rate": 0.5354200988467874,
    "rateDenominator": 607,
    "rateNumerator": 325,
    "year": 2018,
  },
  Object {
    "label": "Jun 2018",
    "locality": "ALL",
    "month": 6,
    "rate": 0.5594936708860759,
    "rateDenominator": 395,
    "rateNumerator": 221,
    "year": 2018,
  },
  Object {
    "label": "Jul 2018",
    "locality": "ALL",
    "month": 7,
    "rate": 0.5414507772020726,
    "rateDenominator": 386,
    "rateNumerator": 209,
    "year": 2018,
  },
  Object {
    "label": "Aug 2018",
    "locality": "ALL",
    "month": 8,
    "rate": 0.5594713656387665,
    "rateDenominator": 227,
    "rateNumerator": 127,
    "year": 2018,
  },
  Object {
    "label": "Sep 2018",
    "locality": "ALL",
    "month": 9,
    "rate": 0.5148514851485149,
    "rateDenominator": 303,
    "rateNumerator": 156,
    "year": 2018,
  },
  Object {
    "label": "Oct 2018",
    "locality": "ALL",
    "month": 10,
    "rate": 0.5517241379310345,
    "rateDenominator": 261,
    "rateNumerator": 144,
    "year": 2018,
  },
  Object {
    "label": "Nov 2018",
    "locality": "ALL",
    "month": 11,
    "rate": 0.49875311720698257,
    "rateDenominator": 401,
    "rateNumerator": 200,
    "year": 2018,
  },
  Object {
    "label": "Dec 2018",
    "locality": "ALL",
    "month": 12,
    "rate": 0.5188284518828452,
    "rateDenominator": 239,
    "rateNumerator": 124,
    "year": 2018,
  },
  Object {
    "label": "Jan 2019",
    "locality": "ALL",
    "month": 1,
    "rate": 0.5625,
    "rateDenominator": 192,
    "rateNumerator": 108,
    "year": 2019,
  },
  Object {
    "label": "Feb 2019",
    "locality": "ALL",
    "month": 2,
    "rate": 0.4907749077490775,
    "rateDenominator": 271,
    "rateNumerator": 133,
    "year": 2019,
  },
  Object {
    "label": "Mar 2019",
    "locality": "ALL",
    "month": 3,
    "rate": 0.5683836589698046,
    "rateDenominator": 563,
    "rateNumerator": 320,
    "year": 2019,
  },
  Object {
    "label": "Apr 2019",
    "locality": "ALL",
    "month": 4,
    "rate": 0.5382932166301969,
    "rateDenominator": 457,
    "rateNumerator": 246,
    "year": 2019,
  },
  Object {
    "label": "May 2019",
    "locality": "ALL",
    "month": 5,
    "rate": 0.5552825552825553,
    "rateDenominator": 407,
    "rateNumerator": 226,
    "year": 2019,
  },
  Object {
    "label": "Jun 2019",
    "locality": "ALL",
    "month": 6,
    "rate": 0.5763546798029556,
    "rateDenominator": 203,
    "rateNumerator": 117,
    "year": 2019,
  },
  Object {
    "label": "Jul 2019",
    "locality": "ALL",
    "month": 7,
    "rate": 0.5150300601202404,
    "rateDenominator": 499,
    "rateNumerator": 257,
    "year": 2019,
  },
  Object {
    "label": "Aug 2019",
    "locality": "ALL",
    "month": 8,
    "rate": 0.5728643216080402,
    "rateDenominator": 597,
    "rateNumerator": 342,
    "year": 2019,
  },
  Object {
    "label": "Sep 2019",
    "locality": "ALL",
    "month": 9,
    "rate": 0.5243362831858407,
    "rateDenominator": 452,
    "rateNumerator": 237,
    "year": 2019,
  },
  Object {
    "label": "Oct 2019",
    "locality": "ALL",
    "month": 10,
    "rate": 0.5720930232558139,
    "rateDenominator": 430,
    "rateNumerator": 246,
    "year": 2019,
  },
  Object {
    "label": "Nov 2019",
    "locality": "ALL",
    "month": 11,
    "rate": 0.5731225296442688,
    "rateDenominator": 506,
    "rateNumerator": 290,
    "year": 2019,
  },
  Object {
    "label": "Dec 2019",
    "locality": "ALL",
    "month": 12,
    "rate": 0.5694444444444444,
    "rateDenominator": 360,
    "rateNumerator": 205,
    "year": 2019,
  },
  Object {
    "label": "Jan 2020",
    "locality": "ALL",
    "month": 1,
    "rate": 0.5425531914893617,
    "rateDenominator": 470,
    "rateNumerator": 255,
    "year": 2020,
  },
  Object {
    "label": "Feb 2020",
    "locality": "ALL",
    "month": 2,
    "rate": 0.5577689243027888,
    "rateDenominator": 251,
    "rateNumerator": 140,
    "year": 2020,
  },
  Object {
    "label": "Mar 2020",
    "locality": "ALL",
    "month": 3,
    "rate": 0.46303501945525294,
    "rateDenominator": 257,
    "rateNumerator": 119,
    "year": 2020,
  },
  Object {
    "label": "Apr 2020",
    "locality": "ALL",
    "month": 4,
    "rate": 0.523121387283237,
    "rateDenominator": 346,
    "rateNumerator": 181,
    "year": 2020,
  },
  Object {
    "label": "May 2020",
    "locality": "ALL",
    "month": 5,
    "rate": 0.537864077669903,
    "rateDenominator": 515,
    "rateNumerator": 277,
    "year": 2020,
  },
  Object {
    "label": "Jun 2020",
    "locality": "ALL",
    "month": 6,
    "rate": 0.4175824175824176,
    "rateDenominator": 182,
    "rateNumerator": 76,
    "year": 2020,
  },
  Object {
    "label": "Jul 2020",
    "locality": "ALL",
    "month": 7,
    "rate": 0.43529411764705883,
    "rateDenominator": 85,
    "rateNumerator": 37,
    "year": 2020,
  },
]
`;

exports[`demographic data ageBucket data 1`] = `
Array [
  Object {
    "label": "<25",
    "rate": 0.49851455733808675,
    "rateDenominator": 1683,
    "rateNumerator": 839,
  },
  Object {
    "label": "25-29",
    "rate": 0.5110088728228722,
    "rateDenominator": 3043,
    "rateNumerator": 1555,
  },
  Object {
    "label": "30-34",
    "rate": 0.5337026777469991,
    "rateDenominator": 1083,
    "rateNumerator": 578,
  },
  Object {
    "label": "35-39",
    "rate": 0.5302879841112215,
    "rateDenominator": 1007,
    "rateNumerator": 534,
  },
  Object {
    "label": "40+",
    "rate": 0.6179775280898876,
    "rateDenominator": 2225,
    "rateNumerator": 1375,
  },
]
`;

exports[`demographic data filtered by locality 1`] = `
Array [
  Object {
    "label": "Total",
    "rate": 0.44340505144995324,
    "rateDenominator": 1069,
    "rateNumerator": 474,
  },
]
`;

exports[`demographic data gender data 1`] = `
Array [
  Object {
    "label": "Male",
    "rate": 0.5458256787099408,
    "rateDenominator": 4899,
    "rateNumerator": 2674,
  },
  Object {
    "label": "Female",
    "rate": 0.5593275137407048,
    "rateDenominator": 3093,
    "rateNumerator": 1730,
  },
]
`;

exports[`demographic data raceOrEthnicity data 1`] = `
Array [
  Object {
    "label": "Native American",
    "rate": 0.3431667670078266,
    "rateDenominator": 1661,
    "rateNumerator": 570,
  },
  Object {
    "label": "Black",
    "rate": 0.5180180180180181,
    "rateDenominator": 666,
    "rateNumerator": 345,
  },
  Object {
    "label": "Hispanic",
    "rate": 0.5048732943469786,
    "rateDenominator": 513,
    "rateNumerator": 259,
  },
  Object {
    "label": "White",
    "rate": 0.5991040318566451,
    "rateDenominator": 10045,
    "rateNumerator": 6018,
  },
  Object {
    "label": "Other",
    "rate": 0.5443037974683544,
    "rateDenominator": 79,
    "rateNumerator": 43,
  },
]
`;

exports[`demographic data total 1`] = `
Array [
  Object {
    "label": "Total",
    "rate": 0.549307774227902,
    "rateDenominator": 9390,
    "rateNumerator": 5158,
  },
]
`;
